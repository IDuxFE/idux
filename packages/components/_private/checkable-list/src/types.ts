/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { VirtualScrollToFn } from '@idux/cdk/scroll'
import type { ExtractInnerPropTypes, ExtractPublicPropTypes, MaybeArray, VKey } from '@idux/cdk/utils'
import type { DefineComponent, HTMLAttributes, PropType } from 'vue'

export interface CheckableListData {
  key?: VKey
  label?: string
  disabled?: boolean
  [key: string]: unknown
}

export interface CheckableListScroll {
  height?: string | number
  fullHeight?: boolean
}

export const checkableListProps = {
  dataSource: Array as PropType<CheckableListData[]>,
  checkable: {
    type: Boolean,
    default: true,
  },
  removable: {
    type: Boolean,
    default: false,
  },
  checked: Function as PropType<(item: CheckableListData) => boolean>,
  customAdditional: { type: Function as PropType<CheckableListCustomAdditional>, default: undefined },
  disabled: Function as PropType<(item: CheckableListData) => boolean>,
  getKey: Function as PropType<(item: CheckableListData) => VKey>,
  labelKey: String,
  virtual: {
    type: Boolean,
    default: false,
  },
  scroll: Object as PropType<CheckableListScroll>,
  onCheckChange: [Function, Array] as PropType<MaybeArray<(item: CheckableListData, checked: boolean) => void>>,
  onRemove: [Function, Array] as PropType<MaybeArray<(item: CheckableListData) => void>>,
  onScroll: [Function, Array] as PropType<MaybeArray<(evt: Event) => void>>,
  onScrolledChange: [Function, Array] as PropType<
    MaybeArray<(startIndex: number, endIndex: number, visibleData: unknown[]) => void>
  >,
  onScrolledBottom: [Function, Array] as PropType<MaybeArray<() => void>>,
} as const

export const checkableListItemProps = {
  checked: {
    type: Boolean,
    default: false,
  },
  checkable: {
    type: Boolean,
    default: true,
  },
  removable: {
    type: Boolean,
    default: false,
  },
  disabled: {
    type: Boolean,
    default: false,
  },
  label: String,
  value: {
    type: [String, Number, Symbol] as PropType<VKey>,
    required: true,
  },
  onCheckChange: [Function, Array] as PropType<MaybeArray<(checked: boolean) => void>>,
  onRemove: [Function, Array] as PropType<MaybeArray<() => void>>,
} as const

export interface CheckableListApi {
  scrollTo: VirtualScrollToFn
}

export type CheckableListProps = ExtractInnerPropTypes<typeof checkableListProps>
export type CheckableListItemProps = ExtractInnerPropTypes<typeof checkableListItemProps>
export type CheckableListPublicProps = ExtractPublicPropTypes<typeof checkableListProps>
export type CheckableListComponent = DefineComponent<
  Omit<HTMLAttributes, keyof CheckableListPublicProps> & CheckableListPublicProps,
  CheckableListApi
>
export type CheckableListInstance = InstanceType<DefineComponent<CheckableListProps, CheckableListApi>>

export type CheckableListCustomAdditional = (options: {
  data: CheckableListData
  index: number
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
}) => Record<string, any> | undefined
