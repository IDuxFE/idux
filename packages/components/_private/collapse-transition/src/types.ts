/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { ExtractInnerPropTypes, ExtractPublicPropTypes } from '@idux/cdk/utils'
import type { DefineComponent, HTMLAttributes } from 'vue'

import { IxPropTypes } from '@idux/cdk/utils'

export type CollapseTransitionMode = 'height' | 'width'

export const collapseTransitionProps = {
  appear: IxPropTypes.bool.def(false),
  name: IxPropTypes.string,
  mode: IxPropTypes.oneOf<CollapseTransitionMode>(['height', 'width']).def('height'),
  onAfterEnter: IxPropTypes.emit(),
  onAfterLeave: IxPropTypes.emit(),
}

export type CollapseTransitionProps = ExtractInnerPropTypes<typeof collapseTransitionProps>
export type CollapseTransitionPublicProps = ExtractPublicPropTypes<typeof collapseTransitionProps>
export type CollapseTransitionComponent = DefineComponent<
  Omit<HTMLAttributes, keyof CollapseTransitionPublicProps> & CollapseTransitionPublicProps
>
export type CollapseTransitionInstance = InstanceType<DefineComponent<CollapseTransitionProps>>
