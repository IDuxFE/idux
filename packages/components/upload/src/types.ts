/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { IxInnerPropTypes, IxPublicPropTypes, VKey } from '@idux/cdk/utils'
import type { ProgressProps } from '@idux/components/progress'
import type { DefineComponent, HTMLAttributes, VNode } from 'vue'

import { IxPropTypes } from '@idux/cdk/utils'

type DataType = Record<string, unknown>
export type UploadRequestHeader = Record<string, string>
export type UploadRequestMethod = 'POST' | 'PUT' | 'PATCH' | 'post' | 'put' | 'patch'
export type UploadRequestStatus = 'loadstart' | 'progress' | 'abort' | 'error' | 'loadend'
export type UploadFileStatus = 'selected' | 'uploading' | 'error' | 'success' | 'abort'
export type UploadListType = 'text' | 'image' | 'imageCard'
export type UploadIconType = 'file' | 'preview' | 'download' | 'remove' | 'retry'
export interface UploadProgressEvent extends ProgressEvent {
  percent?: number
}
export interface UploadRequestError extends Error {
  status?: number
  method?: UploadRequestMethod
  url?: string
}
export interface UploadRawFile extends File {
  uid: VKey
}
export interface UploadFile {
  uid: VKey
  name: string
  raw?: UploadRawFile
  status?: UploadFileStatus
  error?: Error
  errorTip?: string
  thumbUrl?: string
  percent?: number
  response?: Response
}
export interface UploadRequestOption<T = unknown> {
  onProgress?: (event: UploadProgressEvent) => void
  onError?: (event: UploadRequestError | ProgressEvent, body?: T) => void
  onSuccess?: (body: T) => void
  filename: string
  file: UploadRawFile | File
  withCredentials?: boolean
  action: string
  requestHeaders?: UploadRequestHeader
  requestMethod: UploadRequestMethod
  requestData?: DataType
}
export interface UploadRequestChangeOption {
  file: UploadFile
  status: UploadRequestStatus
  response?: Response
  e?: ProgressEvent
}

export const uploadProps = {
  files: IxPropTypes.array<UploadFile>().isRequired,
  accept: IxPropTypes.string,
  action: IxPropTypes.oneOfType([String, IxPropTypes.func<(file: UploadFile) => Promise<string>>()]).isRequired,
  dragable: IxPropTypes.bool,
  directory: IxPropTypes.bool,
  disabled: IxPropTypes.bool,
  maxCount: IxPropTypes.number,
  multiple: IxPropTypes.bool,
  progress: IxPropTypes.object<ProgressProps>(),
  name: IxPropTypes.string,
  customRequest: IxPropTypes.func<(option: UploadRequestOption) => { abort: () => void }>(),
  withCredentials: IxPropTypes.bool,
  requestData: IxPropTypes.oneOfType<DataType | ((file: UploadFile) => DataType | Promise<DataType>)>([
    Object,
    IxPropTypes.func<(file: UploadFile) => DataType | Promise<DataType>>(),
  ]),
  requestHeaders: IxPropTypes.object<UploadRequestHeader>(),
  requestMethod: IxPropTypes.oneOf(['POST', 'PUT', 'PATCH', 'post', 'put', 'patch']),
  'onUpdate:files': IxPropTypes.emit<(fileList: UploadFile[]) => void>(),
  onSelect: IxPropTypes.emit<(file: File[]) => boolean | File[] | Promise<boolean | File[]>>(),
  onBeforeUpload: IxPropTypes.emit<(file: UploadFile) => boolean | UploadFile | Promise<boolean | UploadFile>>(),
  onFileStatusChange: IxPropTypes.emit<(file: UploadFile) => void>(),
  onRequestChange: IxPropTypes.emit<(option: UploadRequestChangeOption) => void>(),
}
export type UploadProps = IxInnerPropTypes<typeof uploadProps>
export type UploadPublicProps = IxPublicPropTypes<typeof uploadProps>
export type UploadComponent = DefineComponent<Omit<HTMLAttributes, keyof UploadPublicProps> & UploadPublicProps>
export type UploadInstance = InstanceType<DefineComponent<UploadProps>>

export const uploadListProps = {
  type: IxPropTypes.oneOf<UploadListType>(['text', 'image', 'imageCard']),
  icon: IxPropTypes.object<Partial<Record<UploadIconType, string | VNode>>>(),
  onDownload: IxPropTypes.emit<(file: UploadFile) => void>(),
  onPreview: IxPropTypes.emit<(file: UploadFile) => void>(),
  onRemove: IxPropTypes.emit<(file: UploadFile) => boolean | Promise<boolean>>(),
  onRetry: IxPropTypes.emit<(file: UploadFile) => void>(),
}
export type UploadListProps = IxInnerPropTypes<typeof uploadListProps>
export type UploadListPublicProps = IxPublicPropTypes<typeof uploadListProps>
export type UploadListComponent = DefineComponent<
  Omit<HTMLAttributes, keyof UploadListPublicProps> & UploadListPublicProps
>
export type UploadListInstance = InstanceType<DefineComponent<UploadListProps>>
