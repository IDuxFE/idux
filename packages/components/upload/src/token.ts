/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { UploadRequest } from './composables/useRequest'
import type { UploadFile, UploadProps } from './types'
import type { ComputedRef, InjectionKey } from 'vue'

export type UploadToken = {
  props: UploadProps
  files: ComputedRef<UploadFile[]>
  onUpdateFiles: (file: UploadFile[]) => void
  setSelectorVisible: (isShow: boolean) => void
} & UploadRequest

export const uploadToken: InjectionKey<UploadToken> = Symbol('UploadToken')
