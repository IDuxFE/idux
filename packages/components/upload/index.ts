/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { UploadComponent, UploadListComponent } from './src/types'

import UploadList from './src/List'
import Upload from './src/Upload'

const IxUpload = Upload as unknown as UploadComponent
const IxUploadList = UploadList as unknown as UploadListComponent

export { IxUpload, IxUploadList }

export type {
  UploadRawFile,
  UploadListType,
  UploadRequestMethod,
  UploadRequestStatus,
  UploadFileStatus,
  UploadProgressEvent,
  UploadFile,
  UploadRequestOption,
  UploadRequestChangeOption,
  UploadIconType,
  UploadInstance,
  UploadComponent,
  UploadPublicProps as UploadProps,
  UploadListInstance,
  UploadListComponent,
  UploadListPublicProps as UploadListProps,
} from './src/types'
