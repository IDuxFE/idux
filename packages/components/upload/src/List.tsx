/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { UploadListProps } from './types'
import type { UploadListConfig } from '@idux/components/config'

import { computed, defineComponent, h, inject, watchEffect } from 'vue'

import { useGlobalConfig } from '@idux/components/config'

import IxUploadImageCardList from './component/ImageCardList'
import IxUploadImageList from './component/ImageList'
import IxUploadTextList from './component/TextList'
import { useSelectorVisible } from './composables/useDisplay'
import { UploadToken, uploadToken } from './token'
import { uploadListProps } from './types'

const cpmMap = {
  text: IxUploadTextList,
  image: IxUploadImageList,
  imageCard: IxUploadImageCardList,
} as const

export default defineComponent({
  name: 'IxUploadList',
  props: uploadListProps,
  setup(props) {
    const config = useGlobalConfig('uploadList')
    const listType = useListType(props, config)
    const { props: uploadProps, setSelectorVisible } = inject(uploadToken, {
      props: {},
      setSelectorVisible: () => {},
    } as unknown as UploadToken)
    const [outerSelector] = useSelectorVisible(uploadProps, listType)

    watchEffect(() => setSelectorVisible(outerSelector.value))

    return () => h(cpmMap[listType.value], { ...props })
  },
})

function useListType(props: UploadListProps, config: UploadListConfig) {
  return computed(() => props.type ?? config.listType)
}
