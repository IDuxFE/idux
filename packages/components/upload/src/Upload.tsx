/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { Ref } from 'vue'

import { defineComponent, provide, ref, shallowRef } from 'vue'

import { useControlledProp } from '@idux/cdk/utils'
import { IxImageViewer } from '@idux/components/image'

import FileSelector from './component/Selector'
import { useCmpClasses } from './composables/useDisplay'
import { useRequest } from './composables/useRequest'
import { uploadToken } from './token'
import { uploadProps } from './types'

export default defineComponent({
  name: 'IxUpload',
  props: uploadProps,
  setup(props, { slots }) {
    const cpmClasses = useCmpClasses()
    const [showSelector, setSelectorVisible] = useShowSelector()
    const [files, onUpdateFiles] = useControlledProp(props, 'files', [])
    const { fileUploading, abort, startUpload, upload } = useRequest(props, files)
    const { viewerVisible, images, setViewerVisible } = useImageViewer()
    provide(uploadToken, {
      props,
      files,
      fileUploading,
      onUpdateFiles,
      abort,
      startUpload,
      upload,
      setViewerVisible,
      setSelectorVisible,
    })

    return () => (
      <div class={cpmClasses.value}>
        {showSelector.value && <FileSelector>{slots.default?.()}</FileSelector>}
        {slots.list?.({ abort, upload })}
        <IxImageViewer v-model:visible={viewerVisible.value} images={images.value}></IxImageViewer>
        <div class={`${cpmClasses.value}-tip`}>{slots.tip?.()}</div>
      </div>
    )
  },
})

function useShowSelector(): [Ref<boolean>, (isShow: boolean) => void] {
  const showSelector = ref(true)

  function setSelectorVisible(isShow: boolean) {
    showSelector.value = isShow
  }

  return [showSelector, setSelectorVisible]
}

function useImageViewer() {
  const viewerVisible = ref(false)
  const images = shallowRef<string[]>([])

  function setViewerVisible(visible: boolean, imageSrc?: string): void {
    images.value = visible && imageSrc ? [imageSrc] : []
    viewerVisible.value = visible
  }

  return {
    viewerVisible,
    images,
    setViewerVisible,
  }
}
