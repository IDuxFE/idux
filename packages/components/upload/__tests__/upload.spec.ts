import type { UploadFile } from '../src/types'
import type { MountingOptions } from '@vue/test-utils'
import type { Ref } from 'vue'

import { DOMWrapper, flushPromises, mount } from '@vue/test-utils'
import { h, ref } from 'vue'

import { renderWork } from '@tests'

import { IxButton } from '@idux/components/button'
import { IxProgress } from '@idux/components/progress'

import UploadFilesListCpm from '../src/List'
import UploadCpm from '../src/Upload'
import { UploadProps } from '../src/types'

const defaultFiles: UploadFile[] = [
  {
    uid: 'test1',
    name: 'idux.svg',
    thumbUrl: '/icons/logo.svg',
  },
  {
    uid: 'error',
    name: 'error.png',
    status: 'error',
    errorTip: 'Upload failed.',
    thumbUrl: '/icons/comp-properties-1.png',
  },
]

const getTestFiles = (count = 1) =>
  new Array(count).fill({}).map((item, index) => new File([`${index}`], `test${index}.png`, { type: 'image/png' }))

// mock input
const triggerInput = async (inputDom: DOMWrapper<HTMLInputElement>, files: File[]) => {
  Object.defineProperty(inputDom.element, 'files', {
    configurable: true,
    writable: true,
    value: files,
  })
  return await inputDom.trigger('change')
}

const uploadMount = (options?: MountingOptions<Partial<UploadProps>>) => {
  const action = '/upload'
  const files: Ref<UploadFile[]> = ref([])
  const { props = {}, slots = {}, ...rest } = options || {}
  const wrapper = mount(UploadCpm, {
    ...rest,
    props: { action, files: files.value, 'onUpdate:files': onUpdateFiles, ...props },
    slots: { list: UploadFilesListCpm, default: h('a', { id: 'upload-test-btn' }), ...slots },
  } as unknown as MountingOptions<UploadProps>)

  function onUpdateFiles(val: UploadFile[]) {
    wrapper.setProps({ files: val })
  }

  return wrapper
}

describe('Upload render', () => {
  let xhrMock: Partial<XMLHttpRequest> = {}

  beforeEach(() => {
    xhrMock = {
      abort: jest.fn(),
      send: jest.fn(),
      open: jest.fn(),
      setRequestHeader: jest.fn(),
      readyState: 4,
      status: 200,
      withCredentials: false,
    }
    jest.spyOn(window, 'XMLHttpRequest').mockImplementation(() => xhrMock as XMLHttpRequest)
  })

  renderWork<UploadProps>(UploadCpm, { props: { action: '/upload', files: [] } })

  test('slots work', async () => {
    const wrapper = uploadMount({
      slots: {
        default: h(IxButton, { id: 'upload-test-trigger' }),
        list: h(
          'div',
          { id: 'upload-test-list' },
          defaultFiles.map(file => h('a', { class: 'upload-list-file' }, file.name)),
        ),
        tip: h('p', { id: 'upload-test-tip' }, 'Tip test'),
      },
    })
    await flushPromises()

    expect(wrapper.find('#upload-test-trigger').exists()).toBe(true)
    expect(wrapper.findAll('.upload-list-file').length).toBe(2)
    expect(wrapper.find('#upload-test-tip').text()).toBe('Tip test')
  })

  test('v-model:files work', async () => {
    const onUpdateFiles = jest.fn()
    const wrapper = uploadMount({
      props: { files: defaultFiles, 'onUpdate:files': onUpdateFiles },
    })
    await flushPromises()

    expect(wrapper.findAll('.ix-upload-file').length).toBe(2)

    await wrapper.setProps({ files: [{ uid: 'files test', name: 'files test' }] })

    expect(wrapper.findAll('.ix-upload-file').length).toBe(1)

    await triggerInput(wrapper.find('.ix-upload-input'), getTestFiles())

    expect(onUpdateFiles).toBeCalled()
  })

  test('accept work', async () => {
    const accept = '.png, image/jpeg'
    const wrapper = uploadMount({ props: { accept } })
    await flushPromises()

    expect(wrapper.find('.ix-upload-input').element.getAttribute('accept')).toBe(accept)

    await triggerInput(wrapper.find('.ix-upload-input'), getTestFiles())

    expect(wrapper.findAll('.ix-upload-file').length).toBe(1) // +1

    await triggerInput(wrapper.find('.ix-upload-input'), [new File(['test'], 'test.jpeg', { type: 'image/jpeg' })])

    expect(wrapper.findAll('.ix-upload-file').length).toBe(2) // +1

    await triggerInput(wrapper.find('.ix-upload-input'), [new File(['test'], 'test.svg', { type: 'image/svg+xml' })])

    expect(wrapper.findAll('.ix-upload-file').length).toBe(2) // illegal
  })

  test('directory work', async () => {
    const wrapper = uploadMount({})
    await flushPromises()

    expect(wrapper.find('.ix-upload-input').element.getAttribute('directory')).toBeFalsy()

    await wrapper.setProps({ directory: true })

    expect(wrapper.find('.ix-upload-input').element.getAttribute('directory')).toBe('directory')
  })

  test('disabled work', async () => {
    const wrapper = uploadMount({})
    await flushPromises()

    expect(wrapper.find('.ix-upload-selector-disabled').exists()).toBeFalsy()

    await wrapper.setProps({ disabled: true })

    expect(wrapper.find('.ix-upload-selector-disabled').exists()).toBeTruthy()
  })

  test('multiple work', async () => {
    const wrapper = uploadMount({})
    await flushPromises()

    expect(wrapper.find('.ix-upload-input').html()).not.toContain('multiple')

    await wrapper.setProps({ multiple: true })

    expect(wrapper.find('.ix-upload-input').html()).toContain('multiple')
  })

  test('maxCount work', async () => {
    const wrapper = uploadMount({ props: { multiple: true, maxCount: 1 } })
    await flushPromises()

    await triggerInput(wrapper.find('.ix-upload-input'), getTestFiles(2))

    expect(wrapper.findAll('.ix-upload-file').length).toBe(1)

    await wrapper.setProps({ maxCount: 2 })
    await triggerInput(wrapper.find('.ix-upload-input'), getTestFiles(3))
    expect(wrapper.findAll('.ix-upload-file').length).toBe(2)

    await wrapper.setProps({ maxCount: 0 })
    await triggerInput(wrapper.find('.ix-upload-input'), getTestFiles(2))
    expect(wrapper.findAll('.ix-upload-file').length).toBe(4) // 2 + 2
  })

  test('dragable work', async () => {
    const wrapper = uploadMount({ props: { dragable: true } })
    await flushPromises()

    const selectorCpm = wrapper.find('.ix-upload-selector-drag')

    expect(selectorCpm.exists()).toBeTruthy()

    await selectorCpm.trigger('dragover')

    expect(selectorCpm.classes()).toContain('ix-upload-selector-dragover')

    await selectorCpm.trigger('dragleave')

    expect(selectorCpm.classes()).not.toContain('ix-upload-selector-dragover')

    await selectorCpm.trigger('drop', { dataTransfer: { files: getTestFiles() } })

    expect(wrapper.findAll('.ix-upload-file').length).toBe(1)
  })

  test('progress work', async () => {
    const wrapper = uploadMount({
      props: {
        files: [
          {
            uid: 'test2',
            name: 'idux.svg',
            status: 'uploading',
            percent: 50,
          },
        ],
      },
    })
    await flushPromises()

    expect(wrapper.findComponent(IxProgress).props()).toMatchObject(expect.objectContaining({ strokeColor: '#20CC94' }))

    await wrapper.setProps({ progress: { strokeColor: { '0%': '#108ee9', '100%': '#87d068' }, strokeWidth: 3 } })

    expect(wrapper.findComponent(IxProgress).props()).toMatchObject(
      expect.objectContaining({
        strokeColor: { '0%': '#108ee9', '100%': '#87d068' },
        strokeWidth: 3,
      }),
    )
  })
})

describe('Upload request', () => {
  let xhrMock: Partial<XMLHttpRequest> = {}

  beforeEach(() => {
    xhrMock = {
      abort: jest.fn(),
      send: jest.fn(),
      open: jest.fn(),
      setRequestHeader: jest.fn(),
      readyState: 4,
      status: 200,
      withCredentials: false,
    }
    jest.spyOn(window, 'XMLHttpRequest').mockImplementation(() => xhrMock as XMLHttpRequest)
  })

  test('action & requestMethod work', async () => {
    const wrapper = uploadMount({
      props: { action: () => Promise.resolve(`/upload/image`) },
    })
    await flushPromises()
    await triggerInput(wrapper.find('.ix-upload-input'), getTestFiles())
    await flushPromises()

    expect(xhrMock.open).toBeCalledWith('post', '/upload/image', true)

    await wrapper.setProps({ action: '/upload/static', requestMethod: 'put' })
    await triggerInput(wrapper.find('.ix-upload-input'), getTestFiles())

    expect(xhrMock.open).toBeCalledWith('put', '/upload/static', true)
  })

  test('name work', async () => {
    let requestParams: FormData | null = null
    xhrMock.send = jest.fn((val: FormData) => (requestParams = val))

    const wrapper = uploadMount({})
    await flushPromises()

    await triggerInput(wrapper.find('.ix-upload-input'), getTestFiles())

    expect(xhrMock.open).toBeCalled()
    expect(xhrMock.send).toBeCalled()
    expect(requestParams!.has('file')).toBe(true)

    await wrapper.setProps({ name: 'testName' })
    await triggerInput(wrapper.find('.ix-upload-input'), getTestFiles())

    expect(requestParams!.has('file')).toBe(false)
    expect(requestParams!.has('testName')).toBe(true)
  })

  test('customRequest work', async () => {
    const customRequest = jest.fn()
    const wrapper = uploadMount({ props: { customRequest: customRequest } })
    await flushPromises()
    await triggerInput(wrapper.find('.ix-upload-input'), getTestFiles())

    expect(xhrMock.open).not.toBeCalled()
    expect(customRequest).toBeCalled()
  })

  test('withCredentials work', async () => {
    const wrapper = uploadMount({})
    await flushPromises()
    await triggerInput(wrapper.find('.ix-upload-input'), getTestFiles())

    expect(xhrMock.withCredentials).toBe(false)

    await wrapper.setProps({ withCredentials: true })
    await triggerInput(wrapper.find('.ix-upload-input'), getTestFiles())

    expect(xhrMock.withCredentials).toBe(true)
  })

  test('requestData work', async () => {
    let resultParams: FormData | null = null
    xhrMock.send = jest.fn((val: FormData) => (resultParams = val))
    const requestData = { testField: Math.random().toString(36).slice(-6) }
    const wrapper = uploadMount({ props: { requestData } })
    await flushPromises()
    await triggerInput(wrapper.find('.ix-upload-input'), getTestFiles())

    expect(resultParams!.get('testField')).toBe(requestData.testField)
  })

  test('requestHeaders work', async () => {
    const resultHeaders: Record<string, string> = {}
    xhrMock.setRequestHeader = jest.fn((name: string, val: string) => (resultHeaders[name] = val))
    const requestHeaders = { testHeader: Math.random().toString(36).slice(-6) }
    const wrapper = uploadMount({ props: { requestHeaders } })
    await flushPromises()
    await triggerInput(wrapper.find('.ix-upload-input'), getTestFiles())

    expect(resultHeaders!.testHeader).toBe(requestHeaders.testHeader)
  })
})

describe('Upload hooks', () => {
  let xhrMock: Partial<XMLHttpRequest> = {}

  beforeEach(() => {
    xhrMock = {
      abort: jest.fn(),
      send: jest.fn(),
      open: jest.fn(),
      setRequestHeader: jest.fn(),
      readyState: 4,
      status: 200,
      withCredentials: false,
    }
    jest.spyOn(window, 'XMLHttpRequest').mockImplementation(() => xhrMock as XMLHttpRequest)
  })

  test('onSelect work', async () => {
    const onSelectNotAllow = jest.fn(() => false)
    const wrapper = uploadMount({ props: { onSelect: onSelectNotAllow } })
    const fileNotAllowSelect = getTestFiles()
    await flushPromises()
    await triggerInput(wrapper.find('.ix-upload-input'), fileNotAllowSelect)

    expect(onSelectNotAllow).toBeCalledWith(fileNotAllowSelect)
    expect(wrapper.findAll('.ix-upload-file').length).toBe(0)

    const onSelectAllow = jest.fn(() => true)
    await wrapper.setProps({ onSelect: onSelectAllow })
    const fileAllowSelect = getTestFiles()
    await flushPromises()
    await triggerInput(wrapper.find('.ix-upload-input'), fileAllowSelect)

    expect(onSelectAllow).toBeCalledWith(fileAllowSelect)
    expect(wrapper.findAll('.ix-upload-file').length).toBe(1)

    const onSelectFileTransform = jest.fn((files: File[]) =>
      files.map(file => Object.assign(file, { transformKey: 'test' })),
    )
    await wrapper.setProps({ onSelect: onSelectFileTransform })
    const fileSelected = getTestFiles()
    await flushPromises()
    await triggerInput(wrapper.find('.ix-upload-input'), fileSelected)

    expect(onSelectFileTransform).toBeCalledWith(
      expect.arrayContaining(fileSelected.map(() => expect.objectContaining({ transformKey: 'test' }))),
    )
    expect(wrapper.findAll('.ix-upload-file').length).toBe(2)
  })

  test('onBeforeUpload work', async () => {
    const onBeforeUploadNotAllow = jest.fn(() => false)
    const wrapper = uploadMount({ props: { onBeforeUpload: onBeforeUploadNotAllow } })
    const fileNotAllowUpload = getTestFiles()
    await flushPromises()
    await triggerInput(wrapper.find('.ix-upload-input'), fileNotAllowUpload)

    expect(onBeforeUploadNotAllow).toBeCalledWith(expect.objectContaining({ raw: fileNotAllowUpload[0] }))
    // Adding to the file list is allowed, but uploading requests are not allowed
    expect(wrapper.findAll('.ix-upload-file').length).toBe(1)
    expect(xhrMock.open).not.toBeCalled()
    ;(xhrMock.open as jest.Mock).mockRestore()

    const onBeforeUploadAllow = jest.fn(() => true)
    await wrapper.setProps({ onBeforeUpload: onBeforeUploadAllow })
    const fileAllowUpload = getTestFiles()
    await flushPromises()
    await triggerInput(wrapper.find('.ix-upload-input'), fileAllowUpload)

    expect(onBeforeUploadAllow).toBeCalledWith(expect.objectContaining({ raw: fileAllowUpload[0] }))
    expect(wrapper.findAll('.ix-upload-file').length).toBe(2)
    expect(xhrMock.open).toBeCalled()
    ;(xhrMock.open as jest.Mock).mockRestore()

    const onBeforeUploadFileTransform = jest.fn((file: UploadFile) => file)
    await wrapper.setProps({ onBeforeUpload: onBeforeUploadFileTransform })
    const fileUpload = getTestFiles()
    await flushPromises()
    await triggerInput(wrapper.find('.ix-upload-input'), fileUpload)

    expect(onBeforeUploadFileTransform).toBeCalledWith(expect.objectContaining({ raw: fileUpload[0] }))
    expect(wrapper.findAll('.ix-upload-file').length).toBe(3)
    expect(xhrMock.open).toBeCalled()

    const onBeforeUploadFilePromise = jest.fn(() => Promise.resolve(true))
    await wrapper.setProps({ onBeforeUpload: onBeforeUploadFilePromise })
    const fileUploadPromise = getTestFiles()
    await flushPromises()
    await triggerInput(wrapper.find('.ix-upload-input'), fileUploadPromise)

    expect(onBeforeUploadFilePromise).toBeCalledWith(expect.objectContaining({ raw: fileUploadPromise[0] }))
    expect(wrapper.findAll('.ix-upload-file').length).toBe(4)
    expect(xhrMock.open).toBeCalled()
  })

  test('onFileStatusChange work', async () => {
    let hasStatus = false
    const onFileStatusChange = jest.fn(
      (file: UploadFile) => (hasStatus = ['selected', 'uploading', 'error', 'success', 'abort'].includes(file.status!)),
    )
    const wrapper = uploadMount({ props: { onFileStatusChange } })
    const fileSelect = getTestFiles()
    await flushPromises()
    await triggerInput(wrapper.find('.ix-upload-input'), fileSelect)

    expect(onFileStatusChange).toBeCalled()
    expect(hasStatus).toBe(true)
  })

  test('onRequestChange work', async () => {
    const onRequestChange = jest.fn()
    const wrapper = uploadMount({ props: { onRequestChange } })
    const fileSelect = getTestFiles()
    await flushPromises()
    await triggerInput(wrapper.find('.ix-upload-input'), fileSelect)

    expect(onRequestChange).toBeCalledWith(expect.objectContaining({ status: 'loadstart' }))

    xhrMock.onload?.call(xhrMock as XMLHttpRequest, null as unknown as ProgressEvent)

    expect(onRequestChange).toBeCalledWith(expect.objectContaining({ status: 'loadend' }))

    Reflect.set(xhrMock, 'status', 400)
    await triggerInput(wrapper.find('.ix-upload-input'), fileSelect)
    xhrMock.onload?.call(xhrMock as XMLHttpRequest, null as unknown as ProgressEvent)

    expect(onRequestChange).toBeCalledWith(expect.objectContaining({ status: 'error' }))
  })
})
