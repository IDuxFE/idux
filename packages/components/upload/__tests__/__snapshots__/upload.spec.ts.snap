// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Upload render render work 1`] = `
"<div class=\\"ix-upload\\">
  <div class=\\"ix-upload-selector\\"><input class=\\"ix-upload-input\\" type=\\"file\\">
    <!---->
  </div>
  <!---->
  <!---->
  <div class=\\"ix-upload-tip\\">
    <!---->
  </div>
</div>"
`;
